contracts/ServiceNodeRewards.sol

```diff
@@ -487,6 +438,76 @@ contract ServiceNodeRewards is Ownable2Step, Pausable, IServiceNodeRewards {
         return _aggregatePubkey;
     }
 
+    /// @notice Add the service node with the specified BLS public key to
+    /// the service node list. EVM revert if the service node already exists.
+    /// @return result The ID allocated for the service node. The service node can then
+    /// be accessed by `_serviceNodes[result]`
+    function serviceNodeAdd(BN256G1.G1Point memory pubkey) internal returns (uint64 result) {
+        // NOTE: Check if the service node already exists
+        // (e.g. <BLS Key> -> <SN> mapping)
+        bytes memory pubkeyBytes = BN256G1.getKeyForG1Point(pubkey);
+        if (serviceNodeIDs[pubkeyBytes] != LIST_SENTINEL)
+            revert BLSPubkeyAlreadyExists(serviceNodeIDs[pubkeyBytes]);
+
+        result             = nextServiceNodeID;
+        nextServiceNodeID += 1;
+        totalNodes        += 1;
+
+        // NOTE: Create service node slot and patch up the slot links.
+        //
+        // The following is the insertion pattern in a doubly-linked list
+        // with sentinel at index 0
+        //
+        // ```c
+        // node->next       = sentinel;
+        // node->prev       = sentinel->prev;
+        // node->next->prev = node;
+        // node->prev->next = node;
+        // ```
+        _serviceNodes[result].next                     = LIST_SENTINEL;
+        _serviceNodes[result].prev                     = _serviceNodes[LIST_SENTINEL].prev;
+        _serviceNodes[_serviceNodes[result].next].prev = result;
+        _serviceNodes[_serviceNodes[result].prev].next = result;
+
+        // NOTE: Assign BLS pubkey
+        _serviceNodes[result].pubkey                   = pubkey;
+
+        // NOTE: Create mapping from <BLS Key> -> <SN Linked List Index>
+        serviceNodeIDs[pubkeyBytes] = result;
+
+        if (totalNodes == 1) {
```

Comment:
I have a question about this, `totalNodes`, how can it go out of sync with the actual length of the list? 

If it can then this wouldn't be safe and I'd switch it back, check if the list is empty before we insert then initialise the aggregate pubkey if it was empty before we inserted. e.g

```
// If the list points to itself, e.g. there are no items.
if (_serviceNodes[LIST_SENTINEL].prev == _serviceNodes[LIST_SENTINEL].next)`
```